File: src/main/java/org/kite9/diagram/visualization/planarization/mgt/MGTPlanarizationImpl.java
package org.kite9.diagram.visualization.planarization.mgt;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;
import java.util.Set;

import org.kite9.diagram.common.BiDirectional;
import org.kite9.diagram.common.elements.edge.Edge;
import org.kite9.diagram.common.elements.edge.PlanarizationEdge;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.Connected;
import org.kite9.diagram.model.Container;
import org.kite9.diagram.model.Diagram;
import org.kite9.diagram.model.DiagramElement;
import org.kite9.diagram.model.position.Dimension2D;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.planarization.Tools;
import org.kite9.diagram.visualization.planarization.ordering.VertexEdgeOrdering;
import org.kite9.diagram.visualization.planarization.rhd.RHDPlanarizationImpl;
import org.kite9.framework.logging.LogicException;

public class MGTPlanarizationImpl extends RHDPlanarizationImpl implements MGTPlanarization {

	public MGTPlanarizationImpl(Diagram d, List<Vertex> vertexOrder, Collection<BiDirectional<Connected>> uninsertedConnections, Map<Container, List<Connected>> containerOrderingMap) {
		super(d, containerOrderingMap);
		this.vertexOrder = vertexOrder;
		this.unmodifiableVO = Collections.unmodifiableList(vertexOrder);
		createVertexIndexMap();
		aboveForwardLinks = new ArrayList<List<PlanarizationEdge>>(vertexOrder.size());
		aboveBackwardLinks = new ArrayList<List<PlanarizationEdge>>(vertexOrder.size());
		belowForwardLinks = new ArrayList<List<PlanarizationEdge>>(vertexOrder.size());
		belowBackwardLinks = new ArrayList<List<PlanarizationEdge>>(vertexOrder.size());
		for (int i = 0; i < vertexOrder.size(); i++) {
			LinkedList<PlanarizationEdge> abl = new LinkedList<>();
			LinkedList<PlanarizationEdge> bbl = new LinkedList<>();
			LinkedList<PlanarizationEdge> afl = new LinkedList<>();
			LinkedList<PlanarizationEdge> bfl = new LinkedList<>();
			aboveBackwardLinks.add(abl);
			aboveForwardLinks.add(afl);
			belowBackwardLinks.add(bbl);
			belowForwardLinks.add(bfl);
		}
		this.uninsertedConnections = uninsertedConnections;
	}

	@Override
	public String toString() {
		if (getFaces().size() == 0) {
			return getTextualRepresentation(null).toString();
		} else {
			return super.toString();
		}
	}

	List<Vertex> vertexOrder;
	List<Vertex> unmodifiableVO;

	private Set<PlanarizationEdge> aboveSet = new LinkedHashSet<PlanarizationEdge>();

	private Set<PlanarizationEdge> belowSet = new LinkedHashSet<PlanarizationEdge>();

	public Set<PlanarizationEdge> getAboveLineEdges() {
		return aboveSet;
	}

	public Set<PlanarizationEdge> getBelowLineEdges() {
		return belowSet;
	}

	public Collection<BiDirectional<Connected>> uninsertedConnections;

	private List<List<PlanarizationEdge>> aboveForwardLinks;
	private List<List<PlanarizationEdge>> aboveBackwardLinks;
	private List<List<PlanarizationEdge>> belowForwardLinks;
	private List<List<PlanarizationEdge>> belowBackwardLinks;

	public List<PlanarizationEdge> getAboveForwardLinks(Vertex v) {
		return aboveForwardLinks.get(getVertexIndex(v));
	}

	public List<PlanarizationEdge> getAboveBackwardLinks(Vertex v) {
		return aboveBackwardLinks.get(getVertexIndex(v));
	}

	public List<PlanarizationEdge> getBelowForwardLinks(Vertex v) {
		return belowForwardLinks.get(getVertexIndex(v));
	}

	public List<PlanarizationEdge> getBelowBackwardLinks(Vertex v) {
		return belowBackwardLinks.get(getVertexIndex(v));
	}

	public List<Vertex> getVertexOrder() {
		return unmodifiableVO;
	}

	public List<Edge> getAllEdges() {
		List<Edge> out = new ArrayList<Edge>();
		out.addAll(aboveSet);
		out.addAll(belowSet);
		return out;
	}

	/**
	 * This does not output crossing edges or on line edges
	 */
	public TextualRepresentation getTextualRepresentation(Set<DiagramElement> highlight) {
		TextualRepresentation tr = new TextualRepresentation();

		// set up vertex positions in the textual rep.
		int voi = 0;
		int lastPos = 0;
		for (Vertex b : vertexOrder) {
			tr.getPositions().put(b, new Dimension2D(lastPos, 0));
			String name = b.getID() + "[" + (voi++) + "]";
			lastPos += name.length() + 2;
		}
		tr.setLength(lastPos);

		voi = 0;
		for (Vertex v : vertexOrder) {
			tr.outputString(0, (int) tr.getPositions().get(v).x(), v.getID() + "[" + (voi++) + "]");
		}

		Map<PlanarizationEdge, Integer> nestings = new HashMap<PlanarizationEdge, Integer>(aboveSet.size() * 2);
		for (PlanarizationEdge e : aboveSet) {
			Vertex from = e.getFrom();
			Vertex to = e.getTo();
			int fromi = (int) tr.getPositions().get(from).x();
			int toi = (int) tr.getPositions().get(to).x();
			int height = getNestings(e, aboveSet, nestings) + 1;
			boolean hl = isPartOf(highlight, e);
			tr.vLine(-1, fromi, -height, hl);
			tr.vLine(-1, toi, -height, hl);
			tr.hLine(-height - 1, fromi + 1, toi - 1, hl);
		}

		nestings = new HashMap<PlanarizationEdge, Integer>(belowSet.size() * 2);
		for (PlanarizationEdge e : belowSet) {
			Vertex from = e.getFrom();
			Vertex to = e.getTo();
			int fromi = (int) tr.getPositions().get(from).x();
			int toi = (int) tr.getPositions().get(to).x();
			int height = getNestings(e, belowSet, nestings) + 1;
			boolean hl = isPartOf(highlight, e);

			tr.vLine(1, fromi, height, hl);
			tr.vLine(1, toi, height, hl);
			tr.hLine(height + 1, fromi + 1, toi - 1, hl);
		}

		return tr;
	}

	private boolean isPartOf(Set<DiagramElement> highlight, Edge e) {
		if (highlight == null) {
			return false;
		}
		for (DiagramElement de : highlight) {
			if (e.isPartOf(de)) {
				return true;
			}
		}
		return false;
	}

	public int getNestings(PlanarizationEdge e, Set<PlanarizationEdge> aboveSet, Map<PlanarizationEdge, Integer> nestCache) {
		Integer out = nestCache.get(e);
		if (out != null) {
			return out;
		}

		int nestings = 0;
		int from = vertexOrder.indexOf(e.getFrom());
		int to = vertexOrder.indexOf(e.getTo());

		for (PlanarizationEdge edge : aboveSet) {
			if (e != edge) {
				int fromi = vertexOrder.indexOf(edge.getFrom());
				int toi = vertexOrder.indexOf(edge.getTo());

				if (((fromi >= from) && (toi <= to)) && (!((fromi == from) && (toi == to)))) {
					nestings = Math.max(nestings, getNestings(edge, aboveSet, nestCache) + 1);
				}
			}
		}

		nestCache.put(e, nestings);

		return nestings;

	}

	public Collection<BiDirectional<Connected>> getUninsertedConnections() {
		return uninsertedConnections;
	}

	public void removeEdge(Edge cross) {
		boolean found = aboveSet.remove(cross);
		int fromvi = getVertexIndex(cross.getFrom());
		int tovi = getVertexIndex(cross.getTo());
		if (found) {
			if (fromvi > -1) {
				aboveForwardLinks.get(fromvi).remove(cross);
			}
			if (tovi > -1) {
				aboveBackwardLinks.get(tovi).remove(cross);
			}
		} else {
			found = belowSet.remove(cross);
			if (found) {
				if (fromvi > -1) {
					belowForwardLinks.get(fromvi).remove(cross);
				}
				if (tovi > -1) {
					belowBackwardLinks.get(tovi).remove(cross);
				}
			}
		}
	}

	private Map<Vertex, Integer> vertexIndex = null;

	private void createVertexIndexMap() {
		vertexIndex = new HashMap<Vertex, Integer>();
		int i = 0;
		for (Vertex vertex : vertexOrder) {
			vertexIndex.put(vertex, i);
			i++;
		}
	}

	public void addVertexToOrder(int i, Vertex insert) {
		vertexOrder.add(i + 1, insert);
		aboveBackwardLinks.add(i + 1, new LinkedList<PlanarizationEdge>());
		aboveForwardLinks.add(i + 1, new LinkedList<PlanarizationEdge>());
		belowBackwardLinks.add(i + 1, new LinkedList<PlanarizationEdge>());
		belowForwardLinks.add(i + 1, new LinkedList<PlanarizationEdge>());
		createVertexIndexMap();
	}

	public int getVertexIndex(Vertex v) {
		Integer i = vertexIndex.get(v);
		if (i == null) {
			return -1;
		} else {
			return i;
		}
	}

	public void removeVertexFromOrder(Vertex v) {
		vertexOrder.remove(v);
		createVertexIndexMap();
	}

	public boolean isAdjacency(Edge edge) {
		Integer val1 = getVertexIndex(edge.getFrom());
		Integer val2 = getVertexIndex(edge.getTo());
		if (val1 == null) {
			throw new LogicException(edge.getFrom() + " not present in diagram");
		}
		if (val2 == null) {
			throw new LogicException(edge.getTo() + " not present in diagram");
		}
		int v1 = val1;
		int v2 = val2;
		return (Math.abs(v1 - v2) <= 1);
	}

	public boolean crosses(Edge edge, boolean above) {
		Vertex v1 = edge.getFrom();
		Vertex v2 = edge.getTo();
		int i1 = getVertexIndex(v1);
		int i2 = getVertexIndex(v2);
		return crosses(i1, i2, above);
	}

	public boolean crosses(float i1, float i2, boolean above) {
		for (Edge edge2 : above ? aboveSet : belowSet) {
			if (crosses(i1, i2, getVertexIndex(edge2.getFrom()), getVertexIndex(edge2.getTo()))) {
				return true;
			}
		}
		return false;
	}

	protected boolean crosses(float i1, float i2, float j1, float j2) {
		if ((i1 == j1) || (i1 == j2) || (i2 == j1) || (i2 == j2))
			return false;

		if (within(i1, j1, j2) != within(i2, j1, j2)) {
			return true;
		}

		return false;
	}

	protected boolean within(float i1, float j1, float j2) {
		float ja = Math.min(j1, j2);
		float jb = Math.max(j1, j2);
		return ((i1 > ja) && (i1 < jb));
	}

	public void addEdge(PlanarizationEdge edge, boolean above, PlanarizationEdge outsideOf) {
		int fromi = getVertexIndex(edge.getFrom());
		int toi = getVertexIndex(edge.getTo());
		if (fromi > toi) {
			edge.reverseDirection();
			int temp = fromi;
			fromi = toi;
			toi = temp;
		}

		if ((outsideOf != null) && (!outsideOf.meets(edge.getFrom()) || !outsideOf.meets(edge.getTo()))) {
			outsideOf = null;
		}

		if (above) {
			aboveSet.add(edge);
			orderedInsert(aboveForwardLinks.get(fromi), edge, outsideOf);
			orderedInsert(aboveBackwardLinks.get(toi), edge, outsideOf);
		} else {
			belowSet.add(edge);
			orderedInsert(belowForwardLinks.get(fromi), edge, outsideOf);
			orderedInsert(belowBackwardLinks.get(toi), edge, outsideOf);
		}
		boolean contradicting = Tools.isUnderlyingContradicting(edge);
		checkEdgeOrdering(edge.getFrom(), edge.getDrawDirectionFrom(edge.getFrom()), contradicting);
		checkEdgeOrdering(edge.getTo(), edge.getDrawDirectionFrom(edge.getTo()), contradicting);

		checkOrderingAround(edge.getFrom());
		checkOrderingAround(edge.getTo());
	}

	private void checkEdgeOrdering(Vertex from, Direction direction, boolean contradicting) {
		// first, fix the vertex ordering
		VertexEdgeOrdering eo = (VertexEdgeOrdering) getEdgeOrderings().get(from);
		if (eo==null) {
			eo = new MGTVertexEdgeOrdering(this, from);
			getEdgeOrderings().put(from, eo);
			eo.addEdgeDirection(direction, contradicting);
		} else {
			eo.addEdgeDirection(direction, contradicting);
			eo.changed();
		}
	}

	private void checkOrderingAround(Vertex from) {
		List<Edge> byQuad = new ArrayList<Edge>();
		List<Edge> c1 = new ArrayList<Edge>(getAboveForwardLinks(from));
		Collections.reverse(c1);
		byQuad.addAll(c1);
		byQuad.addAll(getBelowForwardLinks(from));
		c1 = new ArrayList<Edge>(getBelowBackwardLinks(from));
		Collections.reverse(c1);
		byQuad.addAll(c1);
		byQuad.addAll(getAboveBackwardLinks(from));

		List<PlanarizationEdge> cmp = getEdgeOrderings().get(from).getEdgesAsList();
		int i = byQuad.indexOf(cmp.get(0));
		Collections.rotate(byQuad, -i);

		if (!cmp.equals(byQuad)) {
			throw new LogicException("Collections not same: \n" + cmp + "\n" + byQuad);
		}

	}

	public PlanarizationEdge getFirstEdgeAfterPlanarizationLine(Vertex from, boolean forward, boolean above) {
		PlanarizationEdge outsideOf;
		boolean clockwise = !(above == forward);
		Quadrant q = getQuadrantFor(above, forward);
		List<PlanarizationEdge> edgeSet = getEdgeSetByQuadrant(q, from);
		while (edgeSet.size() == 0) {
			int ord = (q.ordinal() + (clockwise ? -1 : 1) + 4) % 4;
			q = Quadrant.values()[ord];
			edgeSet = getEdgeSetByQuadrant(q, from);
		}

		outsideOf = q.clockwise == clockwise ? edgeSet.get(0) : edgeSet.get(edgeSet.size() - 1);
		return outsideOf;
	}

	private Quadrant getQuadrantFor(boolean above, boolean forward) {
		if (above) {
			return forward ? Quadrant.ABOVE_FORWARD : Quadrant.ABOVE_BACKWARD;
		} else {
			return forward ? Quadrant.BELOW_FORWARD : Quadrant.BELOW_BACKWARD;
		}
	}

	enum Quadrant {
		ABOVE_FORWARD(false), ABOVE_BACKWARD(true), BELOW_BACKWARD(false), BELOW_FORWARD(true);

		private boolean clockwise;

		private Quadrant(boolean clockwise) {
			this.clockwise = clockwise;
		}
	}

	public List<PlanarizationEdge> getEdgeSetByQuadrant(Quadrant quadrant, Vertex v) {
		switch (quadrant) {
		case ABOVE_BACKWARD:
			return getAboveBackwardLinks(v);
		case ABOVE_FORWARD:
			return getAboveForwardLinks(v);
		case BELOW_BACKWARD:
			return getBelowBackwardLinks(v);
		case BELOW_FORWARD:
		default:
			return getBelowForwardLinks(v);
		}
	}

	private void orderedInsert(List<PlanarizationEdge> abl, PlanarizationEdge e, PlanarizationEdge outsideOf) {
		int o1d = edgeSpan(e);
		ListIterator<PlanarizationEdge> li = abl.listIterator();
		while (li.hasNext()) {
			PlanarizationEdge n = li.next();
			int o2d = edgeSpan(n);
			if ((o2d > o1d) || ((o2d == o1d) && (outsideOf == null))) {
				li.previous();
				li.add(e);
				return;
			}
			if ((o2d == o1d) && (outsideOf == n)) {
				li.add(e);
				return;
			}
		}

		li.add(e);
	}

	private int edgeSpan(Edge e) {
		return Math.abs(getVertexIndex(e.getFrom()) - getVertexIndex(e.getTo()));
	}
}

File: a/src/myMain/kotlin/org/kite9/diagram/visualization/planarization/mgt/MGTPlanarizationImpl.kt
package org.kite9.diagram.visualization.planarization.mgt

import org.kite9.diagram.common.BiDirectional
import org.kite9.diagram.common.Collections
import org.kite9.diagram.common.elements.edge.Edge
import org.kite9.diagram.common.elements.edge.PlanarizationEdge
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.*
import org.kite9.diagram.model.position.BasicDimension2D
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.planarization.Tools.Companion.isUnderlyingContradicting
import org.kite9.diagram.visualization.planarization.ordering.VertexEdgeOrdering
import org.kite9.diagram.visualization.planarization.rhd.RHDPlanarizationImpl
import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import kotlin.collections.LinkedHashSet
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

class MGTPlanarizationImpl(
    d: Diagram,
    inputVertexOrder: List<Vertex>,
    inputUninsertedConnections: Collection<BiDirectional<Connected>>,
    containerOrderingMap: Map<Container, List<Connected>>
) : RHDPlanarizationImpl(
    d, containerOrderingMap
), MGTPlanarization {

    override fun toString(): String {
        return if (faces.size == 0) {
            getTextualRepresentation(null).toString()
        } else {
            super.toString()
        }
    }

    override val vertexOrder = inputVertexOrder.toMutableList()
    override val uninsertedConnections = inputUninsertedConnections.toMutableList()
    private val aboveSet: MutableSet<PlanarizationEdge> = LinkedHashSet()
    private val belowSet: MutableSet<PlanarizationEdge> = LinkedHashSet()
    override val aboveLineEdges: Set<PlanarizationEdge>
        get() = aboveSet
    override val belowLineEdges: Set<PlanarizationEdge>
        get() = belowSet

    private val aboveForwardLinks: MutableList<MutableList<PlanarizationEdge>>
    private val aboveBackwardLinks: MutableList<MutableList<PlanarizationEdge>>
    private val belowForwardLinks: MutableList<MutableList<PlanarizationEdge>>
    private val belowBackwardLinks: MutableList<MutableList<PlanarizationEdge>>

    override fun getAboveForwardLinks(v: Vertex): MutableList<PlanarizationEdge> {
        return aboveForwardLinks[getVertexIndex(v)]
    }

    override fun getAboveBackwardLinks(v: Vertex): MutableList<PlanarizationEdge> {
        return aboveBackwardLinks[getVertexIndex(v)]
    }

    override fun getBelowForwardLinks(v: Vertex): MutableList<PlanarizationEdge> {
        return belowForwardLinks[getVertexIndex(v)]
    }

    override fun getBelowBackwardLinks(v: Vertex): MutableList<PlanarizationEdge> {
        return belowBackwardLinks[getVertexIndex(v)]
    }

    override val allEdges: List<Edge>
        get() {
            val out: MutableList<Edge> = ArrayList()
            out.addAll(aboveSet)
            out.addAll(belowSet)
            return out
        }

    /**
     * This does not output crossing edges or on line edges
     */
    fun getTextualRepresentation(highlight: Set<DiagramElement>?): TextualRepresentation {
        val tr = TextualRepresentation()

        // set up vertex positions in the textual rep.
        var voi = 0
        var lastPos = 0
        for (b in vertexOrder) {
            tr.positions[b] = BasicDimension2D(lastPos.toDouble(), 0.0)
            val name = b.getID() + "[" + voi++ + "]"
            lastPos += name.length + 2
        }
        tr.length = lastPos
        voi = 0
        for (v in vertexOrder) {
            tr.outputString(0, tr.positions[v]!!.x().toInt(), v.getID() + "[" + voi++ + "]")
        }
        var nestings: MutableMap<PlanarizationEdge?, Int?> = HashMap(aboveSet.size * 2)
        for (e in aboveSet) {
            val from = e.getFrom()
            val to = e.getTo()
            var fromi = tr.positions[from]!!.x().toInt()
            var toi = tr.positions[to]!!.x().toInt()
            if (fromi > toi) {
                val temp = toi
                toi = fromi
                fromi = temp
            }
            val height = getNestings(e, aboveSet, nestings) + 1
            val hl = isPartOf(highlight, e)
            tr.vLine(-1, fromi, -height, hl)
            tr.vLine(-1, toi, -height, hl)
            tr.hLine(-height - 1, fromi + 1, toi - 1, hl)
        }
        nestings = HashMap(belowSet.size * 2)
        for (e in belowSet) {
            val from = e.getFrom()
            val to = e.getTo()
            var fromi = tr.positions[from]!!.x().toInt()
            var toi = tr.positions[to]!!.x().toInt()
            if (fromi > toi) {
                val temp = toi
                toi = fromi
                fromi = temp
            }
            val height = getNestings(e, belowSet, nestings) + 1
            val hl = isPartOf(highlight, e)
            tr.vLine(1, fromi, height, hl)
            tr.vLine(1, toi, height, hl)
            tr.hLine(height + 1, fromi + 1, toi - 1, hl)
        }
        return tr
    }

    private fun isPartOf(highlight: Set<DiagramElement>?, e: Edge): Boolean {
        if (highlight == null) {
            return false
        }
        for (de in highlight) {
            if (e.isPartOf(de)) {
                return true
            }
        }
        return false
    }

    fun getNestings(
        e: PlanarizationEdge,
        aboveSet: Set<PlanarizationEdge>,
        nestCache: MutableMap<PlanarizationEdge?, Int?>
    ): Int {
        val out = nestCache[e]
        if (out != null) {
            return out
        }
        var nestings = 0
        var from: Int = vertexOrder.indexOf(e.getFrom())
        var to: Int = vertexOrder.indexOf(e.getTo())
        if (from > to) {
            val temp = to
            to = from
            from = temp
        }
        for (edge in aboveSet) {
            if (e !== edge) {
                var fromi: Int = vertexOrder.indexOf(edge.getFrom())
                var toi: Int = vertexOrder.indexOf(edge.getTo())
                if (fromi > toi) {
                    val temp = toi
                    toi = fromi
                    fromi = temp
                }
                if (fromi >= from && toi <= to && !(fromi == from && toi == to)) {
                    nestings = max(nestings, getNestings(edge, aboveSet, nestCache) + 1)
                }
            }
        }
        nestCache[e] = nestings
        return nestings
    }

    override fun removeEdge(cross: Edge) {
        val found = aboveSet.remove(cross) || belowSet.remove(cross)
        val fromvi = getVertexIndex(cross.getFrom())
        val tovi = getVertexIndex(cross.getTo())
        if (fromvi == -1 || tovi == -1) {
            throw LogicException()
        }

        // remove one end
        val found1 = aboveForwardLinks[fromvi].remove(cross) ||
                belowForwardLinks[fromvi].remove(cross) ||
                aboveBackwardLinks[fromvi].remove(cross) ||
                belowBackwardLinks[fromvi].remove(cross)
        val found2 = aboveForwardLinks[tovi].remove(cross) ||
                belowForwardLinks[tovi].remove(cross) ||
                aboveBackwardLinks[tovi].remove(cross) ||
                belowBackwardLinks[tovi].remove(cross)
        if (!found) {
            throw LogicException()
        }
    }

    private val vertexIndex: MutableMap<Vertex, Int> = mutableMapOf()

    private fun createVertexIndexMap() {
        vertexIndex.clear()
        var i = 0
        for (vertex in vertexOrder) {
            vertexIndex[vertex] = i
            i++
        }
    }

    override fun addVertexToOrder(i: Int, insert: Vertex) {
        vertexOrder.add(i + 1, insert)
        aboveBackwardLinks.add(i + 1, mutableListOf())
        aboveForwardLinks.add(i + 1, mutableListOf())
        belowBackwardLinks.add(i + 1, mutableListOf())
        belowForwardLinks.add(i + 1, mutableListOf())
        createVertexIndexMap()
    }

    override fun getVertexIndex(v: Vertex): Int {
        val i = vertexIndex!![v]
        return i ?: -1
    }

    override fun isAdjacency(edge: Edge): Boolean {
        val v1 = getVertexIndex(edge.getFrom())
        val v2 = getVertexIndex(edge.getTo())
        return abs(v1 - v2) <= 1
    }

    override fun crosses(edge: Edge, above: Boolean): Boolean {
        val v1 = edge.getFrom()
        val v2 = edge.getTo()
        val i1 = getVertexIndex(v1)
        val i2 = getVertexIndex(v2)
        return crosses(i1.toFloat(), i2.toFloat(), above)
    }

    override fun crosses(i1: Float, i2: Float, above: Boolean): Boolean {
        for (edge2 in if (above) aboveSet else belowSet) {
            if (crosses(i1, i2, getVertexIndex(edge2.getFrom()).toFloat(), getVertexIndex(edge2.getTo()).toFloat())) {
                return true
            }
        }
        return false
    }

    protected fun crosses(i1: Float, i2: Float, j1: Float, j2: Float): Boolean {
        if (i1 == j1 || i1 == j2 || i2 == j1 || i2 == j2) return false
        return if (within(i1, j1, j2) != within(i2, j1, j2)) {
            true
        } else false
    }

    protected fun within(i1: Float, j1: Float, j2: Float): Boolean {
        val ja = min(j1, j2)
        val jb = max(j1, j2)
        return i1 > ja && i1 < jb
    }

    override fun addEdge(edge: PlanarizationEdge, above: Boolean, outsideOf: PlanarizationEdge?) {
        var outsideOf = outsideOf
        var fromi = getVertexIndex(edge.getFrom())
        var toi = getVertexIndex(edge.getTo())
        if (fromi > toi) {
            val temp = toi
            toi = fromi
            fromi = temp
        }
        if (outsideOf != null && (!outsideOf.meets(edge.getFrom()) || !outsideOf.meets(edge.getTo()))) {
            outsideOf = null
        }
        if (above) {
            aboveSet.add(edge)
            orderedInsert(aboveForwardLinks[fromi], edge, outsideOf)
            orderedInsert(aboveBackwardLinks[toi], edge, outsideOf)
        } else {
            belowSet.add(edge)
            orderedInsert(belowForwardLinks[fromi], edge, outsideOf)
            orderedInsert(belowBackwardLinks[toi], edge, outsideOf)
        }
        val contradicting = isUnderlyingContradicting(edge)
        checkEdgeOrdering(edge.getFrom(), edge.getDrawDirectionFrom(edge.getFrom()), contradicting)
        checkEdgeOrdering(edge.getTo(), edge.getDrawDirectionFrom(edge.getTo()), contradicting)
        checkOrderingAround(edge.getFrom())
        checkOrderingAround(edge.getTo())
    }

    private fun checkEdgeOrdering(from: Vertex, direction: Direction?, contradicting: Boolean) {
        // first, fix the vertex ordering
        var eo = edgeOrderings[from] as VertexEdgeOrdering?
        if (eo == null) {
            eo = MGTVertexEdgeOrdering(this, from)
            edgeOrderings[from] = eo
            eo.addEdgeDirection(direction, contradicting)
        } else {
            eo.addEdgeDirection(direction, contradicting)
            eo.changed()
        }
    }

    private fun checkOrderingAround(from: Vertex) {
        val byQuad: MutableList<Edge> = ArrayList()
        val af: MutableList<PlanarizationEdge> = getAboveForwardLinks(from)
        val bf: MutableList<PlanarizationEdge> = getBelowForwardLinks(from)
        val bb: MutableList<PlanarizationEdge> = getBelowBackwardLinks(from)
        val ab: MutableList<PlanarizationEdge> = getAboveBackwardLinks(from)
        val c1: MutableList<PlanarizationEdge> = af.asReversed()
        byQuad.addAll(c1)
        byQuad.addAll(bf)
        val c2 = bb.asReversed()
        byQuad.addAll(c2)
        byQuad.addAll(ab)
        val cmp = edgeOrderings[from]!!.getEdgesAsList()
        val i = byQuad.indexOf(cmp[0])
        Collections.rotate(byQuad, -i)
        if (cmp != byQuad) {
            throw LogicException("Collections not same: \n$cmp\n$byQuad")
        }
    }

    override fun getFirstEdgeAfterPlanarizationLine(from: Vertex, forward: Boolean, above: Boolean): PlanarizationEdge {
        val outsideOf: PlanarizationEdge
        val clockwise = above != forward
        var q = getQuadrantFor(above, forward)
        var edgeSet = getEdgeSetByQuadrant(q, from)
        while (edgeSet.size == 0) {
            val ord = (q.ordinal + (if (clockwise) -1 else 1) + 4) % 4
            q = Quadrant.values()[ord]
            edgeSet = getEdgeSetByQuadrant(q, from)
        }
        outsideOf = if (q.clockwise == clockwise) edgeSet[0] else edgeSet[edgeSet.size - 1]
        return outsideOf
    }

    private fun getQuadrantFor(above: Boolean, forward: Boolean): Quadrant {
        return if (above) {
            if (forward) Quadrant.ABOVE_FORWARD else Quadrant.ABOVE_BACKWARD
        } else {
            if (forward) Quadrant.BELOW_FORWARD else Quadrant.BELOW_BACKWARD
        }
    }

    enum class Quadrant(val clockwise: Boolean) {
        ABOVE_FORWARD(false), ABOVE_BACKWARD(true), BELOW_BACKWARD(false), BELOW_FORWARD(true);
    }

    fun getEdgeSetByQuadrant(quadrant: Quadrant?, v: Vertex): List<PlanarizationEdge> {
        return when (quadrant) {
            Quadrant.ABOVE_BACKWARD -> getAboveBackwardLinks(v)
            Quadrant.ABOVE_FORWARD -> getAboveForwardLinks(v)
            Quadrant.BELOW_BACKWARD -> getBelowBackwardLinks(v)
            Quadrant.BELOW_FORWARD -> getBelowForwardLinks(v)
            else -> getBelowForwardLinks(v)
        }
    }

    private fun orderedInsert(
        abl: MutableList<PlanarizationEdge>,
        e: PlanarizationEdge,
        outsideOf: PlanarizationEdge?
    ) {
        val o1d = edgeSpan(e)
        val li = abl.listIterator()
        while (li.hasNext()) {
            val n = li.next()
            val o2d = edgeSpan(n)
            if (o2d > o1d || o2d == o1d && outsideOf == null) {
                li.previous()
                li.add(e)
                return
            }
            if (o2d == o1d && outsideOf === n) {
                li.add(e)
                return
            }
        }
        li.add(e)
    }

    private fun edgeSpan(e: Edge): Int {
        return abs(getVertexIndex(e.getFrom()) - getVertexIndex(e.getTo()))
    }

    init {
        createVertexIndexMap()
        aboveForwardLinks = ArrayList(vertexOrder.size)
        aboveBackwardLinks = ArrayList(vertexOrder.size)
        belowForwardLinks = ArrayList(vertexOrder.size)
        belowBackwardLinks = ArrayList(vertexOrder.size)
        for (i in vertexOrder.indices) {
            aboveBackwardLinks.add(mutableListOf())
            aboveForwardLinks.add(mutableListOf())
            belowBackwardLinks.add(mutableListOf())
            belowForwardLinks.add(mutableListOf())
        }
    }
}